### ISSUE TITLE
Extreme Resource Usage During include_role w/ with_items

### ISSUE THREAD

**Issue by 1efty** [ansible/ansible#24891]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
include_role w/ with_items

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```

##### CONFIGURATION
private_role_vars = yes
fact_caching = redis
retry_files_enabled = False

##### OS / ENVIRONMENT
CentOS 7.2
AWS EC2 Instance
t2.medium

##### SUMMARY
Using include_role with a with_items loop causes memory and CPU usage to skyrocket.

##### STEPS TO REPRODUCE

The loop_var looks like this (group_var):
```yaml
virtualhosts:
  - name: name
    state: enabled
    ServerName: name.org
    ssl: yes
    cert: name.crt
    internal_access_only: no
    iframes: yes
    force_https: yes
    password_protect: no
    htaccess: []
    redirect_login_cookie: no
    custom_directives: no
    varnish: yes
    logpath: /var/www/logs
    files: yes
    proxy: yes
    mobile_redirects: no
    balancers:
      - 10.121.67.151
      - 10.121.67.152
      - 10.121.67.153
    bucket: files.name.org
    request_methods:
      - GET
      - POST
      - HEAD
.
.
.
```
All the virtualhosts take up ~800 lines.

The issues occurred here (apache role):
```yaml
- include_role: >
    name=virtualhost
  with_items: "{{ virtualhosts }}"
  loop_control:
    loop_var: virtualhost
  notify: Restart Apache
  tags:
    - apache-virtualhosts

- include_role: >
    name=certificate
  with_items: "{{ virtualhosts | map(attribute='cert') | list }}"
  loop_control:
    loop_var: certificate
  notify: Restart Apache
  tags:
    - apache-certificates
``` 

##### EXPECTED RESULTS
To not use 5GB of memory for two forks. 

##### ACTUAL RESULTS
The playbook runs fine until it reaches the include_role action w\ with_items.  Then the playbook halts and CPU usage is maxed out and memory fills up dipping into about 1GB of swap space.

NOTE: This playbook is only being ran on a single host. 

NOTE: These roles are using copy (synchronize didn't help)

NOTE: These are being ran in a role that has already been included by another playbook.  

i.e:
parent role includes: apache
apache role includes: virtualhost and certificate (both in a with_items) looping through the same virtualhosts variable.

**Comment by krzysztof-magosa** [comment]
I've also hit into that problem in Ansible 2.3 and had to rollback to 2.2.
10 forks could not fit into 8 GB RAM - Ansible reported that dead workers have been found.

**Comment by 1efty** [comment]
I get the same issue on 2.2 and 2.4 (devel)

**Comment by krzysztof-magosa** [comment]
It's really ridiculous ;-)
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 2023 root      20   0 10.378g 8.457g   1048 S  13.7 55.4   7:24.53 ansible-playbook
```

**Comment by ghost** [comment]
I wonder how `with_items:` is currently implemented (can't easily find it... still looking...) my guess is that it'll fully parse the list, then starts working through "it", running out of memory (or just using excessivley much) when it could / would be more efficient (memory wise) to work off a `with_items` list the way a generator would ? memory footprint should be just that of 1 include (or however many are running concurrently / in parallel) .. just some thoughts I had whilst running into the same issue but not even with a role, but simply including another playbook (which is actually quite small) ...

**Comment by rpasche** [comment]
Same here

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
16656 root      20   0 7777864 6,245g   4120 S   0,7 39,8   4:33.69 ansible-playboo
````

**Comment by msieurtoph** [comment]
+1

Any news ?

**Comment by krzysztof-magosa** [comment]
Didn't do any analysis of code, but after moving to Ansible 2.4.1 it looks much better in my case.

**Comment by msieurtoph** [comment]
Tested with ansible 2.4.1 and python 2.7.14
 ... and still get the same problem when using `include_role` and `with_items`

**Comment by kiorky** [comment]
I think #32565 may fix your bug which seems related to #21890

On my fork, you have two branchs for you:

Pristine 2.5+the PR #32565 commits: https://github.com/corpusops/ansible/tree/devel
Pristine 2.4+the PR #32565 commits: https://github.com/corpusops/ansible/tree/a2.4

**Comment by msieurtoph** [comment]
Thx.

I am trying to test @kiorky 's 2.4 fork... I have to chmod +x all the binairies
But I still got this error message when launching: 

```
./ansible-a2.4/bin/ansible-playbook -i inventories/itg1 my-playbook.yml 

Usage: ansible <host-pattern> [options]

Options:
  -a MODULE_ARGS, --args=MODULE_ARGS
                        module arguments
  --ask-vault-pass      ask for vault password
  [....]
    --ask-su-pass       ask for su password (deprecated, use become)
    -K, --ask-become-pass
                        ask for privilege escalation password
ERROR! Missing target hosts
```

It seems like you broke something ...


It works when using "official" 2.4.3.0 ansible tarball.

**Comment by kiorky** [comment]
Nope, something is weird in how you use it, all the test suite is passing, can you install it correctly?
(pip install or python setup.py install)

**Comment by msieurtoph** [comment]
Sorry,
I am not familiar with python environment and I cant get your fork to work.
The VM I am testing on is running several python versions (2.6 and 2.7) AND several ansible versions (1.9 and 2.4), so binaries are scattered all over /usr/bin, /usr/local/bin, /opt/product/, etc ... 

Maybe somebody else with a simplier environment can test your fork.


When I tested the ansible 2.4.1, 2.4.2, 2.4.3, I just had to download the tarball from this url (http://releases.ansible.com/ansible/), unzip, set PYTHONPATH to point the lib path within ... and it worked.
When I tried to download your archive from your github page, unzip, set PYTHONPATH the same way ... it did not. 

Can you provide the same tarball as ansible does in their releases page ?

**Comment by sivel** [comment]
I believe this may be resolved ultimately in devel now as a result of https://github.com/ansible/ansible/pull/38747

Please test and verify.

needs_info

**Comment by sivel** [comment]
I'm going to close this as being resolved.  If you continue experiencing issues, please ensure you test with 2.5.2 or 2.6 (devel) before submitting a new issue.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by msieurtoph** [comment]
@sivel : You officially became my new God with this PR :)
Thank you for it.
